connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetBackingStore.java

```diff
@@ -65,6 +66,13 @@ public interface OffsetBackingStore {
      */
     Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callback);
 
+    /**
+     * Get all the partitions for the specified connector.
+     * @param connectorName the name of the connector whose partitions are to be retrieved
+     * @return set of connector partitions
+     */
+    Set<Map<String, Object>> connectorPartitions(String connectorName);
```

Comment:
There's a couple of things I'd like to note here:

1. As discussed in the [KIP-875 mailing list discussion](https://lists.apache.org/thread/m5bklnh5w4mwr9nbzrmfk0pftpxfjd02), the `OffsetBackingStore` currently only deals with serialized byte buffers because Connect originally had support for pluggable converters for internal topics (which was removed and the`JsonConverter` has been used as the default converter since `2.0.0`).
2. The `OffsetBackingStore` interface was originally designed as a generic key value store. However, the `OffsetStorageWriter` interface is currently the only writer to the offset store interface and it ensures that the keys are of a [certain format](https://github.com/apache/kafka/blob/1560c5bd7e556ca6c0f49934b5ad3542ed6208fb/connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java#L177-L178). The knowledge of this format is required in order to be able to track partitions by connector in the `OffsetBackingStore`.

In either this PR or a follow-up, we should probably refactor the `OffsetBackingStore` interface methods and move all serialization / deserialization from the `OffsetStorageWriter` and `OffsetStorageReader` interfaces into the `OffsetBackingStore` itself. This will make the new `OffsetBackingStore::connectorPartitions` method more in line with the other interface method and also keep the logic for the key format (list of the form `[connectorName, partition]` where `connectorName` is a `string` value and `partition` is a `Map<String, Object>`) in a single central place.

While one could argue that this would be making the `OffsetBackingStore` interface more specific / less generic than it currently is, there aren't any other use cases for the generic offset store currently or in the foreseeable future.

Reply:
Thanks for the detailed writeup! I agree that refactoring the `OffsetBackingStore` interface to handle (de)serialization and connector namespacing would be an improvement, and also that it is not necessary for this PR.

Reply:
I'm glad to hear that we're in agreement here and I've filed this follow-up ticket to make the necessary changes - https://issues.apache.org/jira/browse/KAFKA-14844